### PR TITLE
Feat/swerve drive

### PR DESCRIPTION
**Pull request by danielbrownmsm**
swerve drive support, mostly copied over from last year

TODO:
 * test autonomous
 * fix tele-op.
 * finish unit tests

### CODE REVIEW

File: src/main/cpp/commands/swerve/AutoSwerveCommand.cpp
Comment (jkleiber) [inline src/main/cpp/commands/swerve/AutoSwerveCommand.cpp:26]: what

File: src/main/cpp/commands/swerve/AutoSwerveCommand.cpp
Comment (jkleiber) [inline src/main/cpp/commands/swerve/AutoSwerveCommand.cpp:21]: So is this command supposed to run the whole autonomous? I was assuming we'd do something like send poses and return true in isFinished always, and let the subsystem run the show. Is this a placeholder command?

I'm fine with this as a placeholder as long as we work on this logic for #16

File: src/main/cpp/hardware/HardwareInterface.cpp
Comment (jkleiber) [inline src/main/cpp/hardware/HardwareInterface.cpp:2]: let's rename this file `Hardware.cpp` and the header file `Hardware.h`

File: src/main/cpp/io/SwerveDriveIO.cpp
Comment (jkleiber) [inline src/main/cpp/io/SwerveDriveIO.cpp:44]: Need to OKC_CHECK the AHRS

File: src/main/cpp/io/SwerveDriveIO.cpp
Comment (jkleiber) [inline src/main/cpp/io/SwerveDriveIO.cpp:56]: Whatever is happening below here should definitely be in a separate function lol

File: src/main/cpp/io/SwerveDriveIO.cpp
Comment (jkleiber) [inline src/main/cpp/io/SwerveDriveIO.cpp:92]: OKC_CHECK the AHRS and the encoders before dereferencing with `->`

File: src/main/cpp/io/SwerveDriveIO.cpp
Comment (jkleiber) [inline src/main/cpp/io/SwerveDriveIO.cpp:139]: need to OKC_CHECK the steer motors

File: src/main/cpp/io/SwerveDriveIO.cpp
Comment (jkleiber) [inline src/main/cpp/io/SwerveDriveIO.cpp:173]: Need to OKC_CHECK these encoders

File: src/main/deploy/parameters.toml
Comment (jkleiber) [inline src/main/deploy/parameters.toml:2]: change title for 2023

File: src/main/include/SwerveModule.h
Comment (jkleiber) [inline src/main/include/SwerveModule.h:82]: insert comment about trailing underscores

File: src/main/include/commands/swerve/AutoSwerveCommand.h
Comment (jkleiber) [inline src/main/include/commands/swerve/AutoSwerveCommand.h:18]: I think we need more than one command, but we can discuss later

File: src/main/include/hardware/ActuatorInterface.h
Comment (jkleiber) [inline src/main/include/hardware/ActuatorInterface.h:23]: actuators_t, or whatever is in `feat/hardware`

File: src/main/include/hardware/ActuatorInterface.h
Comment (jkleiber) [inline src/main/include/hardware/ActuatorInterface.h:36]: Actuators

File: src/main/include/hardware/HardwareInterface.h
Comment (jkleiber) [inline src/main/include/hardware/HardwareInterface.h:20]: Hardware

File: src/main/include/hardware/SensorInterface.h
Comment (jkleiber) [inline src/main/include/hardware/SensorInterface.h:38]: Sensors

File: src/main/include/subsystems/SwerveDrive.h
Comment (jkleiber) [inline src/main/include/subsystems/SwerveDrive.h:77]: trailing underscores on private vars

File: src/main/include/ui/UserInterface.h
Comment (jkleiber) [inline src/main/include/ui/UserInterface.h:40]: We can move these pointers (and below) to a struct and then put them in the SwerveDrive subsystem instead

Comment (jkleiber) [reply src/main/include/ui/UserInterface.h:40]: I'm fine with not doing this until later though

File: src/test/cpp/subsystems/swerve_drive_test.cpp
Comment (jkleiber) [inline src/test/cpp/subsystems/swerve_drive_test.cpp:165]: This passes because you aren't checking anything with ASSERT or EXPECT

File: src/main/cpp/subsystems/SwerveDrive.cpp
Comment (jkleiber) [inline src/main/cpp/subsystems/SwerveDrive.cpp:16]: parameterize these pls

File: src/main/cpp/subsystems/SwerveDrive.cpp
Comment (jkleiber) [inline src/main/cpp/subsystems/SwerveDrive.cpp:100]: Checks should be before use

File: src/main/cpp/subsystems/SwerveDrive.cpp
Comment (jkleiber) [inline src/main/cpp/subsystems/SwerveDrive.cpp:96]: Checks should be before use

File: src/main/cpp/subsystems/SwerveDrive.cpp
Comment (jkleiber) [inline src/main/cpp/subsystems/SwerveDrive.cpp:144]: need a check on swerve_odometry

File: src/main/cpp/subsystems/SwerveDrive.cpp
Comment (jkleiber) [inline src/main/cpp/subsystems/SwerveDrive.cpp:135]: can we make this a function

File: src/main/cpp/subsystems/SwerveDrive.cpp
Comment (jkleiber) [inline src/main/cpp/subsystems/SwerveDrive.cpp:141]: can we make this a function

File: src/main/cpp/subsystems/SwerveDrive.cpp
Comment (jkleiber) [inline src/main/cpp/subsystems/SwerveDrive.cpp:243]: need OKC_CHECKs on the modules

File: src/main/cpp/subsystems/SwerveDrive.cpp
Comment (jkleiber) [inline src/main/cpp/subsystems/SwerveDrive.cpp:250]: Need an OKC_CHECK on the interface_

File: src/main/cpp/subsystems/SwerveDrive.cpp
Comment (jkleiber) [inline src/main/cpp/subsystems/SwerveDrive.cpp:279]: This could totally be run in Periodic and then the command would just set something like `mode = PoseFollowAuto` and then this just runs

Comment (jkleiber) [inline src/main/cpp/subsystems/SwerveDrive.cpp:279]: modules need OKC_CHECK

Comment (jkleiber) [reply src/main/cpp/subsystems/SwerveDrive.cpp:279]: So does the interface_

File: src/main/cpp/subsystems/SwerveDrive.cpp
Comment (jkleiber) [inline src/main/cpp/subsystems/SwerveDrive.cpp:363]: Do `*heading = interface_->imu_yaw` instead

File: src/main/cpp/subsystems/SwerveDrive.cpp
Comment (jkleiber) [inline src/main/cpp/subsystems/SwerveDrive.cpp:394]: Are these functions useful for swerve?

File: src/main/cpp/subsystems/SwerveDrive.cpp
Comment (jkleiber) [inline src/main/cpp/subsystems/SwerveDrive.cpp:408]: need okc check for `at`

File: src/main/cpp/subsystems/SwerveDrive.cpp
Comment (jkleiber) [inline src/main/cpp/subsystems/SwerveDrive.cpp:424]: OKC_CHECK these first

File: src/main/cpp/subsystems/SwerveDrive.cpp
Comment (jkleiber) [inline src/main/cpp/subsystems/SwerveDrive.cpp:470]: OKC_CHECK these

File: src/main/cpp/subsystems/SwerveDrive.cpp
Comment (jkleiber) [inline src/main/cpp/subsystems/SwerveDrive.cpp:476]: OKC_CHECK shuffleboard pointers

Comment (jkleiber) [reply src/main/cpp/subsystems/SwerveDrive.cpp:476]: also check drive modules for nullptr

File: src/main/cpp/subsystems/SwerveDrive.cpp
Comment (jkleiber) [inline src/main/cpp/subsystems/SwerveDrive.cpp:554]: okc check interface for nullptr

**Review by jkleiber — COMMENTED** [review]
A lot of comments

**Review by jkleiber — COMMENTED** [review]
A lot of comments

File: src/main/cpp/commands/swerve/TeleOpSwerveCommand.cpp
Comment (noahzemlin) [inline src/main/cpp/commands/swerve/TeleOpSwerveCommand.cpp:27]: There doesn't need to be two negations on drive_power

File: src/main/cpp/commands/swerve/AutoSwerveCommand.cpp
Comment (danielbrownmsm) [reply src/main/cpp/commands/swerve/AutoSwerveCommand.cpp:48]: then how do you do auto? does the command scheduler keep running motors and subsystems once the autonomous command ends?

Comment (jkleiber) [reply src/main/cpp/commands/swerve/AutoSwerveCommand.cpp:48]: Subsystem periodic() is always called, so our commands should be "set and forget" and we let Periodic set values over and over like you describe

File: src/test/cpp/subsystems/swerve_drive_test.cpp
Comment (danielbrownmsm) [reply src/test/cpp/subsystems/swerve_drive_test.cpp:165]: but I figured if it threw an exception somewhere it should fail the test right?

File: src/main/cpp/subsystems/SwerveDrive.cpp
Comment (danielbrownmsm) [reply src/main/cpp/subsystems/SwerveDrive.cpp:418]: I kinda just had them there on shuffleboard to make sure things were actually working. They're somewhat useful for swerve, but individual NetworkTableEntries for each module would be best

File: src/test/cpp/subsystems/swerve_drive_test.cpp
Comment (jkleiber) [reply src/test/cpp/subsystems/swerve_drive_test.cpp:165]: we dont throw exceptions though - we only return true early if a check or call fails. so the function just exits gracefully.

if you had `ASSERT_TRUE(Periodic());` you coukd catch the failure

**Review by jkleiber — COMMENTED** [review]
Once all comments are addressed and this branch gets rebased, I think this is ready to merge. Let's hold off any swerve changes until we get this merged because it's getting hard to review

File: src/main/cpp/SwerveModule.cpp
Comment (jkleiber) [inline src/main/cpp/SwerveModule.cpp:26]: Setting these default to zero may result in divide by zero if params fail to load. Consider a different default like 1.

**Review by jkleiber — COMMENTED** [review]
One comment. While you fix that I'll go try to figure out if I can get the navX library to work on Linux x86

**Review by jkleiber — APPROVED** [review]
Sure why not

**Pull request merged**
